### DISCUSSION TITLE
🚀  Edge caching

### DISCUSSION THREAD

**Discussion by spencerwooo** [spencerwooo/onedrive-vercel-index#407] [Announcements]
## 🚀  Performance improvements with edge caching

Folder listing, thumbnail, search, etc., are now cached with Vercel's edge function caching. You should experience extremely faster load speed for these routes, especially folder listings. _(Note: the thumbnails themselves are not cached to follow Vercel's fair use policy, only the URL pointing to the thumbnails are cached.)_

**The cache is valid for 60 seconds on the edge.** - This means you won't see your newly added files for at least 60 seconds after your files are done uploading.

If you want to manually purge the cache of your site, you can do so by triggering a redeployment manually.

> The cache is automatically purged upon a new deployment being created. If you ever need to invalidate the CDN cache, you can always re-deploy. Further reading: https://vercel.com/docs/concepts/edge-network/caching.

## 🛠 Technical details

Relevant headers:

```
Cache-Control: max-age=0, s-maxage=60, stale-while-revalidate
```

You can check for response headers `x-vercel-cache` to see whether your request is responsed by a cache hit (fast 🐰), or fresh revalidated data (slow 🐢).

### Cache miss

Most often, no cache is present at the edge when someone first visits a route. This is presented as _cache miss_, where the response header `x-vercel-cache` is set to `MISS`. In this scenario, fresh data is requested from the upstream API, and you will experience a slow load speed (compared to other scenarios).

![CleanShot 2022-02-10 at 23 42 37](https://user-images.githubusercontent.com/32114380/153442855-1ff6ac92-f9c5-4dce-9e7e-dbb863fcc993.png)

### Cache hit

After the route is requested, Vercel will cache the function response at the edge. When a second request on the same route is triggered, Vercel will respond with the cache instead of requesting data from the upstream for a second time. This is presented as _cache hit_, where the response header `x-vercel-cache` is set to `HIT`. You will experience an instant response.

![CleanShot 2022-02-10 at 23 40 17](https://user-images.githubusercontent.com/32114380/153442575-22394604-5ff8-4452-be1d-1969dd1ed867.png)

### Cache stale

If the cache is present for longer than 60 seconds, on the request triggered, Vercel will respond with the stale cache, setting `x-vercel-cache` to `STALE`, and perform a new request to the upstream API. In this case, the following response is guaranteed to be fresh, and Vercel will update the cache on the edge accordingly. You will experience an instant response for both requests.

![CleanShot 2022-02-10 at 23 41 27](https://user-images.githubusercontent.com/32114380/153442621-18a8583c-48d3-4187-9936-5a3b2b7f473a.png)

**Comment by Android-KitKat** [comment]
👀 好耶！缓存终于来了。
`Cache stale` 的情况，也就是说刷新两次才会看到最新内容吗？

**Reply by spencerwooo** [reply]
https://vercel.com/docs/concepts/functions/edge-caching#stale-while-revalidate

> Which tells our CDN: serve from cache, but update it, if requested after ~~1 second~~ 10 minutes.

也就是 10 分钟后的请求，如果发现 stale 了，先返回 cache 的 response，同时更新 cache，这样下一次 response 是最新内容。

**Comment by Android-KitKat** [comment]
我发现个小问题，如果 `STALE` 的缓存在超过一个小时后被访问，那么 `@microsoft.graph.downloadUrl` 是无效的。
然后 在线浏览 和 下载按钮 就会出错，需要再次刷新才能使用。Σ( ￣□￣||)<

所以 `stale-while-revalidate` 设置为 31天，[驱动器项目](https://docs.microsoft.com/zh-cn/graph/api/resources/driveitem?view=graph-rest-1.0) 可能仍会存在，但其中的 `@microsoft.graph.downloadUrl` 一定不可用。

|柱|ω・`) 我觉得也许还能再优化一下？

**Reply by spencerwooo** [reply]
确实，`@microsoft.graph.downloadUrl` 的有效时间是 1 小时是吧？应该也设置 cache 有效时间。

**Reply by Android-KitKat** [reply]
> 确实，`@microsoft.graph.downloadUrl` 的有效时间是 1 小时是吧？应该也设置 cache 有效时间。

对，文档写是 1小时。
